src/Compiler/Utilities/range.fs

```diff
@@ -566,3 +697,61 @@ module Range =
                     | None -> mkRange file (mkPos 1 0) (mkPos 1 80)
         with _ ->
             mkRange file (mkPos 1 0) (mkPos 1 80)
+
+module internal FileContent =
+    let private fileContentDict = ConcurrentDictionary<string, string array>()
+    let private newlineMarkDict = ConcurrentDictionary<string, string>()
+
+    let readFileContents (fileNames: string list) =
+        for fileName in fileNames do
+            if FileSystem.FileExistsShim fileName then
+                use fileStream = FileSystem.OpenFileForReadShim(fileName)
+                let text = fileStream.ReadAllText()
+
+                let newlineMark =
+                    if text.IndexOf('\n') > -1 && text.IndexOf('\r') = -1 then
+                        "\n"
+                    else
+                        "\r\n"
+
+                newlineMarkDict[fileName] <- newlineMark
+                fileContentDict[fileName] <- text.Split([| newlineMark |], StringSplitOptions.None)
+
+    type IFileContentGetLine =
+        abstract GetLine: fileName: string -> line: int -> string
+        abstract GetLineNewLineMark: fileName: string -> string
+
+    let mutable getLineDynamic =
+        { new IFileContentGetLine with
+            member this.GetLine (fileName: string) (line: int) : string =
+                match fileContentDict.TryGetValue fileName with
+                | true, lines when lines.Length > line -> lines[line - 1]
+                | _ -> String.Empty
+
+            member this.GetLineNewLineMark(fileName: string) : string =
+                match newlineMarkDict.TryGetValue fileName with
+                | true, res -> res
+                | _ -> String.Empty
+        }
+
+    let getCodeText (m: range) =
```

Comment:
I wonder whether it would be possible to use [`ISourceText.GetSubTextString`](https://github.com/dotnet/fsharp/blob/0a715a580500d30c2e6aa53e84513e22958b2616/src/Compiler/Facilities/prim-lexing.fsi#L23-L24) instead of all of this. I think the `ISourceText` for a given file will usually already be cached when this functionality is needed.

I think in theory the source text is available on `cenv.tcSink.CurrentSink.Value.CurrentSourceText.Value`, but maybe there's a better way to get it, or a better way to bring it in scope for this change.

Reply:
![图片](https://github.com/user-attachments/assets/4ce3f13d-f4db-47c9-ba19-4fa5db3cbfb6)

It seems that this cannot work under `dotnet fsi`

Reply:
Maybe it would make sense to pass it in, then, as is done when checking format strings:

https://github.com/dotnet/fsharp/blob/0a715a580500d30c2e6aa53e84513e22958b2616/src/Compiler/Checking/NameResolution.fs#L1777-L1779

https://github.com/dotnet/fsharp/blob/0a715a580500d30c2e6aa53e84513e22958b2616/src/Compiler/Checking/CheckFormatStrings.fs#L67-L73

Reply:
> Maybe it would make sense to pass it in, then, as is done when checking format strings:

I guess maybe that's the same source text as in the sink... But still, maybe there's somewhere else we could put it.

Reply:
> > Maybe it would make sense to pass it in, then, as is done when checking format strings:
> 
> I guess maybe that's the same source text as in the sink... But still, maybe there's somewhere else we could put it (or a function/object for lazily getting an expression's source).